pkg/icingadb/driver.go

```diff
@@ -33,24 +46,28 @@ func (c RetryConnector) Connect(ctx context.Context) (driver.Conn, error) {
 		ctx,
 		func(ctx context.Context) (err error) {
 			conn, err = c.Connector.Connect(ctx)
+			if err == nil && c.initConn != nil {
+				err = c.initConn(ctx, conn)
```

Comment:
If c.Connector.Connect() succeeds, but c.initConn() doesn't, shouldn't we close conn explicitly?

Reply:
We shouldn't actually have retries around the `initConn` part, I just put it there after https://github.com/Icinga/icingadb/pull/711#discussion_r1531629225 without thinking about it thoroughly. If there are retryable errors, `c.Connector.Connect()` should have retried them and given up at that point, so there is no need for additional retries. Anyway, the connection should definitely be closed!

Reply:
> We shouldn't actually have retries around the initConn part

If a connection is established but for some reason, `initConn()` encounters a connection error (connection reset by peer for example), why shouldn't it be retried with a new connection?

> If there are retryable errors, `c.Connector.Connect()` should have retried them and given up at that point, so there is no need for additional retries.

Isn't `c.Connector.Connect()` simply the connect function of `go-sql-driver/mysql` or `lib/pq` which doesn't have any special retry handling?

Reply:
> If a connection is established but for some reason, initConn() encounters a connection error (connection reset by peer for example), why shouldn't it be retried with a new connection?

If `c.Connector.Connect()` succeeds and `initConn()` fails, we would have to close every successfully established connection after each retry and waste time just because of `initConn`.

> Isn't c.Connector.Connect() simply the connect function of go-sql-driver/mysql or lib/pq which doesn't have any special retry handling?

I mean our retry handler around the `Connect()` call by that.

Reply:
> > If a connection is established but for some reason, initConn() encounters a connection error (connection reset by peer for example), why shouldn't it be retried with a new connection?
> 
> If `c.Connector.Connect()` succeeds and `initConn()` fails, we would have to close every successfully established connection after each retry and waste time just because of `initConn`.

I think I start to understand what you're thinking here. Is it that if `initConn` would receive some kind of "not ready yet, try again later" error, which would then result in the connection being thrown away even though it would probably get into a working state later?

So yes, there are situations were connection might be closed unnecessarily.

However, consider the scenario when a new connection is established about the same time when the database server is restarted. So `Connect()` could still have worked perfectly fine, however, `initConn()` might receive some error like "server is shutting down" or "connection closed". If you don't retry `initConn()` or retry it separately from `Connect()`, this would report an error while it would have worked perfectly fine if it reattempted with a fresh connection.

Now unless you try to distinguish these situations (we got to #698, so I don't think we want to try this here), the safe option is to close the connection and try again on a new one, even if not strictly necessary in some situations.